### PR TITLE
Improve mdast functionality with mystjs and myst-spec

### PR DESCRIPTION
**Pull request by fwkoch**
- [x] @fwkoch I think we can bring your other changes for table --> mdast import in here
- [x] citations get close to something here https://github.com/executablebooks/myst-spec/issues/21
  - [x] I think only implement the group and label/identifier for now.
- [x] table figure numbering maybe isn't working in the mdast export
- [x] table *not* in a figure needs to be caught (recent changes broke this)
- [x] table rowspan, colspan, header working and tested
- [x] mdast localize localizeImageSrc, ... etc.
- [x] reactive nodes
- [x] double check that everything can export/import to mdast!
  - [x] footnotes
  - [x] time
  - [x] iframe
  - [x] aside
  - [x] numbered on math
  - [x] mentions

Later on Tables:
- tables widths

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -57,6 +58,7 @@
     "@types/js-yaml": "^4.0.5",
     "@types/jsdom": "^16.2.12",
     "@types/markdown-it": "^12.2.3",
+    "@types/mdast": "^3.0.10",
```
Comment (rowanc1) [inline package.json:61]: Do we need this?

File: src/nodes/basic.ts [hunk 2]
```diff
@@ -16,54 +30,106 @@ export const docComment: NodeSpec = {
...
   toDOM() {
     return ['br'];
   },
+  attrsFromMdastToken: () => ({}),
+  toMyst: (): Break => ({ type: 'break' }),
+};
+
+export type ListAttrs = {
+  order: number | null;
 };
```
Comment (rowanc1) [inline src/nodes/basic.ts:95]: Can this come from myst-spec?

File: src/nodes/basic.ts [hunk 3]
```diff
@@ -16,54 +30,106 @@ export const docComment: NodeSpec = {
...
   NodeGroups.block,
-);
+) as OrderedMap<MyNodeSpec<any, any>>;
+
+export type OrderedListAttrs = {
+  order: number;
+};
+
+export const ordered_list = listNodes.get('ordered_list') as MyNodeSpec<OrderedListAttrs, List>;
+ordered_list.attrsFromMdastToken = (token: GenericNode) => ({ order: token.start || 1 });
```
Comment (rowanc1) [inline src/nodes/basic.ts:108]: Can probably type this?

Comment (rowanc1) [reply src/nodes/basic.ts:108]: I think it is start no?
https://executablebooks.github.io/myst-spec/myst.schema.html#list

Confused about the `order` prop.

File: src/nodes/footnote.ts [hunk 4]
```diff
@@ -12,6 +12,8 @@ const footnote: MyNodeSpec<any> = {
   atom: true,
   toDOM: () => ['span', { class: 'footnote' }, 0],
   parseDOM: [{ tag: 'span.footnote' }],
+  attrsFromMdastToken: () => ({}),
+  toMyst: () => ({}),
```
Comment (rowanc1) [inline src/nodes/footnote.ts:16]: Footnotes!

File: src/nodes/mention.ts [hunk 5]
```diff
@@ -28,6 +28,8 @@ function createMentionNodeSpecs() {
         },
       },
     ],
+    attrsFromMdastToken: () => ({ label: '', user: '' }),
+    toMyst: () => ({}),
```
Comment (rowanc1) [inline src/nodes/mention.ts:32]: TODO: mention

File: src/nodes/time.ts [hunk 6]
```diff
@@ -53,6 +54,8 @@ const time: MyNodeSpec<Attrs> = {
       },
     },
   ],
+  attrsFromMdastToken: () => ({ datetime: null }),
+  toMyst: (props): Text => ({ type: 'text', value: formatDatetime(props.datetime).f }),
```
Comment (rowanc1) [inline src/nodes/time.ts:58]: We can probably introduce a time prop here that is the ISO formatted datetime string.

File: src/nodes/utils.ts [hunk 7]
```diff
@@ -128,3 +135,55 @@ export function getColumnWidths(node: Node<any>) {
...
+  }
+  state.write(`\`\`\`{mdast} ${mdastId}`);
+  state.ensureNewLine();
+  state.write('```');
+  state.closeBlock(node);
+  return true;
+}
+
+// TODO: this is directly from mystjs - we should export from there instead
+export function normalizeLabel(
```
Comment (rowanc1) [inline src/nodes/utils.ts:180]: Can we follow through on this todo!

File: src/parse/mdast/index.ts [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+  tokens: GenericNode[],
+) => {
+  name: nodeNames | markNames | ignoreNames;
+  children?: GenericNode[] | string;
+};
+
+// type MdastHandler = {
+//   block: string;
+//   getAttrs?: (node: GenericNode) => Record<string, any>;
+// };
```
Comment (rowanc1) [inline src/parse/mdast/index.ts:143]:
```suggestion
```

File: src/parse/mdast/index.ts [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+const handlers: Record<string, TokenHandler> = {
```
Comment (rowanc1) [inline src/parse/mdast/index.ts:145]: Are these missing a whole bunch of things? time, reactive nodes ....

**Review by rowanc1 — CHANGES_REQUESTED** [review]
Took a more detailed look through the PR. Tons in here! :)

File: src/nodes/basic.ts (hunk 3 above)
Comment (fwkoch) [reply src/nodes/basic.ts:108]: `order` is on the schema / prosemirror side - this is the number the list starts at: https://prosemirror.net/docs/ref/#schema-list.orderedList. It's `start` on the myst side.

I think maybe there is a problem a few lines down though...
```typescript
ordered_list.toMyst = 
  ...
  start: props.start || undefined,
  ...
});
```
should have `start: props.order || undefined` I think.

File: src/nodes/basic.ts (hunk 2 above)
Comment (rowanc1) [reply src/nodes/basic.ts:95]: This will be deleted!

**Review by fwkoch — COMMENTED** [review]
Good stuff - last few commits look good to me.

**Pull request merged**